pkg/webhook/podunavailablebudget/validating/pub_create_update_handler.go

```diff
@@ -0,0 +1,147 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validating
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"reflect"
+
+	policyv1alpha1 "github.com/openkruise/kruise/apis/policy/v1alpha1"
+	"github.com/openkruise/kruise/pkg/util"
+
+	metavalidation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	appsvalidation "k8s.io/kubernetes/pkg/apis/apps/validation"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// PodUnavailableBudgetCreateUpdateHandler handles PodUnavailableBudget
+type PodUnavailableBudgetCreateUpdateHandler struct {
+	// To use the client, you need to do the following:
+	// - uncomment it
+	// - import sigs.k8s.io/controller-runtime/pkg/client
+	// - uncomment the InjectClient method at the bottom of this file.
+	Client client.Client
+
+	// Decoder decodes objects
+	Decoder *admission.Decoder
+}
+
+var _ admission.Handler = &PodUnavailableBudgetCreateUpdateHandler{}
+
+// Handle handles admission requests.
+func (h *PodUnavailableBudgetCreateUpdateHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	obj := &policyv1alpha1.PodUnavailableBudget{}
+
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	allErrs := h.validatingSidecarSetFn(obj)
+	if len(allErrs) != 0 {
+		return admission.Errored(http.StatusInternalServerError, allErrs.ToAggregate())
+	}
+	return admission.ValidationResponse(true, "")
+}
+
+func (h *PodUnavailableBudgetCreateUpdateHandler) validatingSidecarSetFn(obj *policyv1alpha1.PodUnavailableBudget) field.ErrorList {
+	//validate Pub.Spec
+	allErrs := validatePodUnavailableBudgetSpec(obj, field.NewPath("spec"))
+
+	//validate whether pub is in conflict with others
+	pubList := &policyv1alpha1.PodUnavailableBudgetList{}
+	if err := h.Client.List(context.TODO(), pubList, &client.ListOptions{Namespace: obj.Namespace}); err != nil {
+		allErrs = append(allErrs, field.InternalError(field.NewPath(""), fmt.Errorf("query other podUnavailableBudget failed, err: %v", err)))
+	} else {
+		allErrs = append(allErrs, validatePubConflict(obj, pubList.Items, field.NewPath("spec"))...)
+	}
+	return allErrs
+}
+
+func validatePodUnavailableBudgetSpec(obj *policyv1alpha1.PodUnavailableBudget, fldPath *field.Path) field.ErrorList {
+	spec := &obj.Spec
+	allErrs := field.ErrorList{}
+
+	if spec.Selector == nil && spec.TargetReference == nil {
```

Comment:
should also validate `spec.Selector != nil && spec.TargetReference != nil` for they are mutually exclusive